fl4health/server/base_server.py

```diff
@@ -163,6 +164,90 @@ def poll_clients_for_sample_counts(self, timeout: Optional[float]) -> List[int]:
 
         return sample_counts
 
+    def _split_metrics(
+        self, results: List[Tuple[ClientProxy, EvaluateRes]]
+    ) -> Tuple[List[Tuple[ClientProxy, EvaluateRes]], List[Tuple[ClientProxy, EvaluateRes]]]:
+        val_results = []
+        test_results = []
+
+        for client_proxy, eval_res in results:
+            val_metrics = {k: v for k, v in eval_res.metrics.items() if not k.startswith("test - ")}
+            test_metrics = {k: v for k, v in eval_res.metrics.items() if k.startswith("test - ")}
+
+            # Remove loss and num_examples from test_metrics if they exist
+            test_loss = float(test_metrics.pop("test - loss"))
+            test_num_examples = int(test_metrics.pop("test - num_examples"))
+
+            non_test_eval_res = EvaluateRes(eval_res.status, eval_res.loss, eval_res.num_examples, val_metrics)
+            test_eval_res = EvaluateRes(eval_res.status, test_loss, test_num_examples, test_metrics)
+
+            val_results.append((client_proxy, non_test_eval_res))
+            test_results.append((client_proxy, test_eval_res))
+
+        return val_results, test_results
```

Comment:
I mocked up a slight change to this workflow with an assertion and avoiding the need to do any key checking in the `_evaluate_round` implementation (see below). Let me know if you see any issues with the potential changes.

Reply:
`    def _unpack_metrics(
        self, results: List[Tuple[ClientProxy, EvaluateRes]]
    ) -> Tuple[List[Tuple[ClientProxy, EvaluateRes]], List[Tuple[ClientProxy, EvaluateRes]]]:
        val_results = []
        test_results = []

        for client_proxy, eval_res in results:
            val_metrics = {k: v for k, v in eval_res.metrics.items() if not k.startswith("test - ")}
            test_metrics = {k: v for k, v in eval_res.metrics.items() if k.startswith("test - ")}

            if len(test_metrics) > 0:
                assert "test - loss" in test_metrics and "test - num_examples" in test_metrics, (
                    "test - loss and test - num_examples keys must be present "
                    "in test_metrics dictionary for aggregation"
                )
                # Remove loss and num_examples from test_metrics if they exist
                test_loss = float(test_metrics.pop("test - loss"))
                test_num_examples = int(test_metrics.pop("test - num_examples"))
                test_eval_res = EvaluateRes(eval_res.status, test_loss, test_num_examples, test_metrics)
                test_results.append((client_proxy, test_eval_res))

            val_eval_res = EvaluateRes(eval_res.status, eval_res.loss, eval_res.num_examples, val_metrics)
            val_results.append((client_proxy, val_eval_res))

        return val_results, test_results`

Reply:
```
    def _unpack_metrics(
        self, results: List[Tuple[ClientProxy, EvaluateRes]]
    ) -> Tuple[List[Tuple[ClientProxy, EvaluateRes]], List[Tuple[ClientProxy, EvaluateRes]]]:
        val_results = []
        test_results = []

        for client_proxy, eval_res in results:
            val_metrics = {k: v for k, v in eval_res.metrics.items() if not k.startswith("test - ")}
            test_metrics = {k: v for k, v in eval_res.metrics.items() if k.startswith("test - ")}

            if len(test_metrics) > 0:
                assert "test - loss" in test_metrics and "test - num_examples" in test_metrics, (
                    "test - loss and test - num_examples keys must be present "
                    "in test_metrics dictionary for aggregation"
                )
                # Remove loss and num_examples from test_metrics if they exist
                test_loss = float(test_metrics.pop("test - loss"))
                test_num_examples = int(test_metrics.pop("test - num_examples"))
                test_eval_res = EvaluateRes(eval_res.status, test_loss, test_num_examples, test_metrics)
                test_results.append((client_proxy, test_eval_res))

            val_eval_res = EvaluateRes(eval_res.status, eval_res.loss, eval_res.num_examples, val_metrics)
            val_results.append((client_proxy, val_eval_res))

        return val_results, test_results
```

Reply:
I think this make mroe sense will add the test- related changes here